modules/dnn/src/layers/reorg_layer.cpp

```diff
@@ -0,0 +1,139 @@
+//     and/or other materials provided with the distribution.
+//
+//   * The name of the copyright holders may not be used to endorse or promote products
+//     derived from this software without specific prior written permission.
+//
+// This software is provided by the copyright holders and contributors "as is" and
+// any express or implied warranties, including, but not limited to, the implied
+// warranties of merchantability and fitness for a particular purpose are disclaimed.
+// In no event shall the Intel Corporation or contributors be liable for any direct,
+// indirect, incidental, special, exemplary, or consequential damages
+// (including, but not limited to, procurement of substitute goods or services;
+// loss of use, data, or profits; or business interruption) however caused
+// and on any theory of liability, whether in contract, strict liability,
+// or tort (including negligence or otherwise) arising in any way out of
+// the use of this software, even if advised of the possibility of such damage.
+//
+//M*/
+
+#include "../precomp.hpp"
+#include <opencv2/dnn/shape_utils.hpp>
+#include <opencv2/dnn/all_layers.hpp>
+#include <iostream>
+
+namespace cv
+{
+namespace dnn
+{
+
+class ReorgLayerImpl : public ReorgLayer
+{
+    int reorgStride;
+public:
+
+    ReorgLayerImpl(const LayerParams& params)
+    {
+        setParamsFrom(params);
+
+        reorgStride = params.get<int>("reorg_stride", 2);
+        CV_Assert(reorgStride > 0);
+    }
+
+    bool getMemoryShapes(const std::vector<MatShape> &inputs,
+                         const int requiredOutputs,
+                         std::vector<MatShape> &outputs,
+                         std::vector<MatShape> &internals) const
+    {
+        CV_Assert(inputs.size() > 0);
+        outputs = std::vector<MatShape>(inputs.size(), shape(
+            inputs[0][0],
+            inputs[0][1] * reorgStride * reorgStride,
+            inputs[0][2] / reorgStride,
+            inputs[0][3] / reorgStride));
+
+        CV_Assert(outputs[0][0] > 0 && outputs[0][1] > 0 && outputs[0][2] > 0 && outputs[0][3] > 0);
+
+        return false;
+    }
+
+    virtual bool supportBackend(int backendId)
+    {
+        return backendId == DNN_BACKEND_DEFAULT;
+    }
+    void forward(std::vector<Mat*> &inputs, std::vector<Mat> &outputs, std::vector<Mat> &internals)
+    {
+        CV_TRACE_FUNCTION();
+        CV_TRACE_ARG_VALUE(name, "name", name.c_str());
+
+        for (size_t i = 0; i < inputs.size(); i++)
+        {
+            Mat srcBlob = *inputs[i];
+            MatShape inputShape = shape(srcBlob), outShape = shape(outputs[i]);
+            float *dstData = outputs[0].ptr<float>();
+            const float *srcData = srcBlob.ptr<float>();
+
+            int channels = inputShape[1], height = inputShape[2], width = inputShape[3];
+
+            int out_c = channels / (reorgStride*reorgStride);
+
+            for (int k = 0; k < channels; ++k) {
```

Comment:
Please make it more clear: iterate over output dimensions and map them to input ones.

Reply:
Most likely there was a logical mistake in the original version of Darknet in the reorg layer: https://github.com/pjreddie/darknet/blob/master/src/blas.c#L9

It works as I described if called `reorg(input, output, out_w, out_h, _out_c, ...);` https://github.com/opencv/opencv/pull/9705#discussion_r141307309

But in original Darknet version function is called in this way `reorg(input, output, in_w, in_h, ...);`, so the one-to-one correspondence of the input and output parameters is preserved, but very strange permutations occur.

But because original Darknet works with this implementation of reorg and all models trained using it, then we can't fix logical mistake.

Why the author has not found and corrected this error? I think:
* Perhaps this logical error does not spoil the detection accuracy, so it was not detected. 
* Theoretically, we can assume that this error even increased the accuracy - so the author found it and left it.

----

> iterate over output dimensions and map them to input ones.

So I can implement it in a such way, 
but it works correctly only **`if (in_w % 2 == 0 && in_h % 2 == 0 && in_c % 4 == 0)`** : http://coliru.stacked-crooked.com/a/b962d20938362d4f
```
void reorg_my(const float*const srcData,  float *const dstData, int width, int height, int channels, int reorgStride)
{
	int outChannels = channels * reorgStride * reorgStride;
	int outHeight = height / reorgStride;
	int outWidth = width / reorgStride;

	for (int y = 0; y < outHeight; ++y) {
		for (int x = 0; x < outWidth; ++x) {
			for (int c = 0; c < outChannels; ++c) {
				int out_index = x + outWidth*(y + outHeight*c);

				int step = c / channels;
				int x_offset = step % reorgStride;
				int y_offset = reorgStride * ((step / reorgStride) % reorgStride);

				int in_x = x * reorgStride + x_offset;
				
				int out_seq_y = y + c*outHeight;
				int in_intermediate_y = out_seq_y*2 - out_seq_y%2;
				in_intermediate_y = in_intermediate_y % (channels*height);
				int in_c = in_intermediate_y / height;
				int in_y = in_intermediate_y % height + y_offset;
						
				int in_index = in_x + width*(in_y + height*in_c);
				dstData[out_index] = srcData[in_index];
			}
		}
	}
}
```

Reply:
> Most likely there was made a logical mistake in the original version of Darknet in the reorg layer: https://github.com/pjreddie/darknet/blob/master/src/blas.c#L9
> 
> It works as I described if called `reorg(input, output, out_w, out_h, _out_c, ...);` [#9705 (comment)](https://github.com/opencv/opencv/pull/9705#discussion_r141307309)
> 
> But in the original Darknet version function is called in this way `reorg(input, output, in_w, in_h, in_c, ...);`, so the one-to-one correspondence of the input and output parameters is preserved, but very strange permutations occur.
> 
> But because original Darknet works with this implementation of reorg and all models trained using it, then we can't fix this logical mistake.
> 
> Why the author has not found and corrected this error? I think:
> 
>     * Perhaps this logical error does not spoil the detection accuracy, so it was not detected.
> 
>     * Theoretically, we can assume that this error even increased the accuracy - so the author found it and left it.
> 
> 
> > iterate over output dimensions and map them to input ones.
> 
> So I can implement it in a such way,
> but it works correctly only **`if (in_w % 2 == 0 && in_h % 2 == 0 && in_c % 4 == 0)`** : http://coliru.stacked-crooked.com/a/b962d20938362d4f
> 
> ```
> void reorg_my(const float*const srcData,  float *const dstData, int width, int height, int channels, int reorgStride)
> {
> 	int outChannels = channels * reorgStride * reorgStride;
> 	int outHeight = height / reorgStride;
> 	int outWidth = width / reorgStride;
> 
> 	for (int y = 0; y < outHeight; ++y) {
> 		for (int x = 0; x < outWidth; ++x) {
> 			for (int c = 0; c < outChannels; ++c) {
> 				int out_index = x + outWidth*(y + outHeight*c);
> 
> 				int step = c / channels;
> 				int x_offset = step % reorgStride;
> 				int y_offset = reorgStride * ((step / reorgStride) % reorgStride);
> 
> 				int in_x = x * reorgStride + x_offset;
> 				
> 				int out_seq_y = y + c*outHeight;
> 				int in_intermediate_y = out_seq_y*2 - out_seq_y%2;
> 				in_intermediate_y = in_intermediate_y % (channels*height);
> 				int in_c = in_intermediate_y / height;
> 				int in_y = in_intermediate_y % height + y_offset;
> 						
> 				int in_index = in_x + width*(in_y + height*in_c);
> 				dstData[out_index] = srcData[in_index];
> 			}
> 		}
> 	}
> }
> ```

Is there a GPU version of reorg？